### ISSUE TITLE
test

### ISSUE THREAD

**Issue by ywfwyht** [ywfwyht/keylane#1]
`'''
Date: 2023-03-24 08:42:16
Author: yang_haitao
LastEditors: yanghaitao yang_haitao@leapmotor.com
LastEditTime: 2023-03-27 07:27:54
FilePath: /K-Lane/home/work_dir/work/keylane/models/scheduler/CosineLRWarmup.py
'''
""" Cosine Scheduler

Cosine LR schedule with warmup, cycle/restarts, noise, k-decay.

Hacked together by / Copyright 2021 Ross Wightman
"""
import logging
import math
import numpy as np
import torch
import timm
from torch.optim.lr_scheduler import CosineAnnealingLR, _LRScheduler

_logger = logging.getLogger(__name__)

class CosineLRWarmup(_LRScheduler):
    """
    Cosine decay with restarts.
    This is described in the paper https://arxiv.org/abs/1608.03983.

    k-decay option based on `k-decay: A New Method For Learning Rate Schedule` - https://arxiv.org/abs/2004.05909
    """

    def __init__(self, optimizer, T_max, eta_min=1.0e-6, last_epoch=-1, verbose=False,
                 warmup_steps=2):
        super(CosineLRWarmup, self).__init__(optimizer, last_epoch, verbose)
        self.warmup_steps = warmup_steps
        self.eta_min = eta_min
        self.T_max = T_max
        if warmup_steps:
            self.warup_factors = [
                (base_lr - eta_min) / (self.warmup_steps * self.warmup_steps)
                for base_lr in self.base_lrs
            ]

    def get_lr(self):
        if not self._get_lr_called_within_step:
            _logger.warn("To get the last learning rate computed by the scheduler, "
                          "please use `get_last_lr()`.", UserWarning)
            
        if self.last_epoch == 0:
            return [1.0e-8]
        if hasattr(self,'warmup_steps'):
            if self.last_epoch < self.warmup_steps:
                return [self.eta_min + (warmup_factor*self.last_epoch*self.last_epoch)
                        for warmup_factor in self.warup_factors]
            else:
                return [self.eta_min + (base_lr - self.eta_min) *
                        (1 + math.cos(math.pi * (self.last_epoch - self.warmup_steps) / (self.T_max - self.warmup_steps)))*0.5
                        for base_lr in self.base_lrs]
        else:
            return [self.eta_min + (base_lr - self.eta_min) *
                (1 + math.cos(math.pi * self.last_epoch / self.T_max)) / 2
                for base_lr in self.base_lrs]
   

if __name__ == "__main__":
    import matplotlib.pyplot as plt
    lr_list = []
    # all_pretrained_models_available = timm.list_models(pretrained=True)
    # print(all_pretrained_models_available)
    model = timm.create_model('resnet18')
    optimizer = torch.optim.Adam(
        model.parameters(), 
        lr=0.002, 
        betas=(0.9, 0.999), 
        eps=1e-8)
    lr_scheduler = CosineLRWarmup(optimizer,
                        T_max=100,
                        eta_min=1.0e-8,
                        last_epoch=-1,
                        warmup_steps=5)
    for epoch in range(100):
        
        lr_list.append(optimizer.param_groups[0]['lr'])
        optimizer.step()
        lr_scheduler.step()

    plt.figure(figsize=(10, 6))
    plt.plot(range(100), lr_list, color = 'r')
    # plt.text(0, lr_list[0], str(lr_list[0]))
    # plt.text(100, lr_list[-1], str(lr_list[-1]))
    plt.show()
`
